grpcbp/server_middlewares.go

```diff
@@ -0,0 +1,77 @@
+package grpcbp
+
+import (
+	"context"
+
+	"github.com/reddit/baseplate.go/ecinterface"
+	"github.com/reddit/baseplate.go/log"
+	"github.com/reddit/baseplate.go/tracing"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+func InjectServerSpanInterceptor() grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (_ interface{}, err error) {
+		ctx, span := StartSpanFromGRPCContext(ctx, info.FullMethod)
+		defer func() {
+			span.FinishWithOptions(tracing.FinishOptions{
+				Ctx: ctx,
+				Err: err,
+			}.Convert())
+		}()
+		return handler(ctx, req)
+	}
+}
+
+func InjectEdgeContextInterceptor(impl ecinterface.Interface) grpc.UnaryServerInterceptor {
+	if impl == nil {
+		impl = ecinterface.Get()
+	}
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (_ interface{}, err error) {
+		ctx = InitializeEdgeContext(ctx, impl)
+		return handler(ctx, req)
+	}
+}
+
+func InitializeEdgeContext(ctx context.Context, impl ecinterface.Interface) context.Context {
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return ctx
+	}
+	values, ok := md[HeaderEdgeRequest]
+	if !ok {
+		return ctx
+	}
+
+	ctx, err := impl.HeaderToContext(ctx, values[0])
+	if err != nil {
+		log.Error("Error while parsing EdgeRequestContext: " + err.Error())
+	}
+	return ctx
+}
+
+func StartSpanFromGRPCContext(ctx context.Context, name string) (context.Context, *tracing.Span) {
+	var headers tracing.Headers
+	var sampled bool
+
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return tracing.StartSpanFromHeaders(ctx, name, headers)
+	}
+
+	if values, ok := md[HeaderTracingTrace]; ok {
```

Comment:
I _think_ the recommended way is to use `Get` over accessing the map directly? so this should be something like:
```
if values := md.Get(HeaderTracingTrace); len(values) > 0 {
```

looking at the doc of `MD`, it does mention that it will automatically convert uppercase keys into lowercase, so accessing the uppercase keys directly might result in always getting not ok here.